Create analytic cohort: Update to accept multiple cohorts and stack resulting data together

jalavery opened the pull request:
This is a big one - happy to set up time to discuss these changes if that would be helpful. Not at all urgent, we can discuss at the spring hackathon if that's easier for both of you.

For PR submission:
What changes are involved in this pull request? 
- Updated `create_analytic_cohort` to allow the function to take multiple cohorts from `pull_data_synapse()` and stack the data together. This update is based on stacking the data together and then running through the selection criteria (as opposed to running through the criteria and then stacking together - attempted that in PR #149 and ran into roadblocks)
- This required coercing some variable types to character so that the data could be stacked together across releases
- Created a lookup table for institutions available by cohort. This made checking the correct institutions easier, and will be helpful as we move into phase 2 data releases with more institutions. 
- Updated checks for histology to account for the case where only brca data is supplied vs brca + non-brca data since histology variables are different (will have to do similar when RCC is released)
- Changed all `message()` calls to `cli::cli_inform()` so that all are printed (with message, if multiple criteria triggered a message, only the last was printed)

For consideration/discussion:
- Should we add more tests related to supplying multiple cohorts?
- When we return stacked data for multiple cohorts, we need a variable to indicate the cohort because some patients appear in multiple cancer cohorts. A few decision points related to this:
1. Do we want to indicate this via a single variable combining the cohort + data release (e.g., NSCLC_v2.0) or two variables, one for cohort, one for data release? For now, used a single variable. 
2. Do we want to return the variable(s) always (even if only a single cancer cohort is supplied to create_analytic_cohort) or only when multiple cancer cohorts are specified? For now, only returning when multiple cohorts are supplied (when only 1 cohort is supplied, creating this variable based on the name of the list is harder and will require a different approach)
3. Does including a cohort variable on the genomic data result in any downstream issues if trying to use gnomeR functions to process the genomic data?

Other notes:
- The tests all pass, but sometimes the message is printed after the expect_message() and sometimes it's not. Not sure why that is or if it matters?
- `drug_regimen_sunburst` can still only handle a single cohort at a time. Not sure that we want to update, but if we do, should be a separate PR. 

Is there a GitHub issue corresponding to this pull request? If so, please provide link. #112 (see also PR #149)

Checklist for PR Reviewer
- [x] Make sure all updates from main branch are pulled to branch issuing pull request
- [x] Confirm package dependencies are installed by running `renv::install()`
- [ ] For bug corrections, check that unit test was added
- [x] Document changes from this pull request in `NEWS.md` file
- [ ] `devtools::spell_check()` runs with no spelling errors in documentation
- [x] R CMD Check runs without errors, warnings, and notes
- [ ] Run `pkgdown::build_site()`. Check the R console for errors, and review the rendered website. If there are errors returned, try running `pkgdown::build_site(new_process = FALSE)` for better error messaging.
- [ ] Code coverage is suitable for any new functions/features. Review coverage with `covr::report()`. Before you run, begin in a fresh R session without any packages loaded and set `Sys.setenv(NOT_CRAN="true")`.
- [ ] Increment the version number using `usethis::use_version(which = "dev")`
- [ ] Approve and merge pull request

slb2240 commented:
Meeting 3/11/2025:
- Karissa and Sammi to both review the code
- Sammi to review & update tests
- Karissa to review Github actions

jalavery commented:
5/15/2025

- In older versions of purrr, the list_transpose function only returns items present in the first list. E.g., when combining NSCLC 3.1-consortium (no RT) and CRC 2.0-public (RT), RT data is not returned. This isn't an issue in purrr 1.0.4+. Tests do not fail on older versions of purrr. Considerations: (1) Update required version of purrr - downside is that it's the most recent version,  may introduce issues with package versions for users (2) Program around it with Hadley's work-around, but he notes that it increases run time. Will go with #1 since current CRAN version. SB to add two tests related to this (checking stacked rows/cols matches sum of individual, all datasets present in any cohort are returned).
- Add messaging for combining cohorts where datasets are not available for all cohorts (e.g., SV/fusions, RT, TM) (JL)
- Combine cohort_institution table with synapse_versions table (JL)
- Potential future hackathon task: Make the cohort_institution and release_dates into a CSV file that's read in instead of hard coding in R (add to list for issue filed about new cohorts)
- Another potential hackathon task to make more modular functions w/in create_analytic_cohort (e.g., cohort_xx dataset)

jalavery commented:
5/29/2025
**Done since previous meeting**
- Added tests for checking rows for all possible combos of data releases and check for names of each instrument (SB)
- Added messaging for combining cohorts where datasets are not available for all cohorts (e.g., SV/fusions, RT, TM) (JL) 
- Removed .data global binding - was added after running goodpractice::gp but is not current practice w/ the latest version of tidyselect (JL)
- Made tests more modular (KW)

**In progress**
- In older versions of purrr, the list_transpose function only returns items present in the first list. E.g., when combining NSCLC 3.1-consortium (no RT) and CRC 2.0-public (RT), RT data is not returned. This isn't an issue in purrr 1.0.4+. Tests do not fail on older versions of purrr. Considerations: (1) Update required version of purrr - downside is that it's the most recent version,  may introduce issues with package versions for users (2) Program around it with Hadley's work-around, but he notes that it increases run time. Will go with option 1 since current CRAN version. 
- SB will add check for number of columns (won't be the sum of cols or the min # cols, have to look at intersection + differences)
- KW continuing making tests more modular. Will move tests for create analytic cohort into separate file depending on the level of access permissions
- JL to look into modify_header warnings removing 'update'

**Hackathon**
- Potential future hackathon task: Make the cohort_institution and release_dates into a CSV file that's read in instead of hard coding in R (add to list for issue filed about new cohorts)
- Another potential hackathon task to make more modular functions w/in create_analytic_cohort (e.g., cohort_xx dataset)
- Combine cohort_institution table with synapse_versions table
- In create_analytic_cohort() tests, may consider putting tests for all data releases in a separate test file that's run intermittantly, while keeping main tests running on current versions of each cancer cohort
